File: AuctionMarket/src/main/java/com/portfolio/auctionmarket/domain/categories/entity/Category.java
package com.portfolio.auctionmarket.domain.categories.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.SQLRestriction;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity(name = "categories")
@Table(name = "categories", indexes = {
        @Index(name = "idx_category_parent", columnList = "parent_id")
})
@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE category_id = ?")
@SQLRestriction("is_deleted = false")
@Builder
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "category_id")
    private Long categoryId;

    @Column(name = "category")
    private String category;

    @Column(name = "path")
    private String path;

    @Column(name = "is_deleted")
    private String isDeleted;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private Category parent;

    @Builder.Default
    @OneToMany(mappedBy = "parent", cascade = CascadeType.ALL)
    private List<Category> children = new ArrayList<>();

    public void addChildrenCategory(Category children) {
        this.children.add(children);
        children.setParent(this);
    }
}

File: AuctionMarket/auction-domain/src/main/kotlin/org/example/auctioncommon/domain/categories/entity/Category.kt
package org.example.auctioncommon.domain.categories.entity

import jakarta.persistence.*
import org.hibernate.annotations.SQLDelete
import org.hibernate.annotations.SQLRestriction

@Entity(name = "categories")
@Table(
    name = "categories",
    indexes = [Index(name = "idx_category_parent", columnList = "parent_id")]
)
@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE category_id = ?")
@SQLRestriction("is_deleted = false")
class Category(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "category_id")
    var categoryId: Long? = null,

    @Column(name = "category")
    var category: String,

    @Column(name = "path")
    var path: String? = null,

    @Column(name = "is_deleted")
    var isDeleted: String? = "false",

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    var parent: Category? = null,

    @OneToMany(mappedBy = "parent", cascade = [CascadeType.ALL])
    var children: MutableList<Category?> = mutableListOf()

) {

    fun addChildrenCategory(children: Category) {
        this.children.add(children)
        children.parent = this
    }
}